distributed/stealing.py

```diff
@@ -331,6 +346,19 @@ def balance(self):
                             continue
 
                         i += 1
+
+                        # in case the key has resource_restrictions,
+                        # set i to next worker providing them
+                        if key in self.scheduler.resource_restrictions.keys():
+                            resource_i = self.next_resource_i(
+                                idle,
+                                self.scheduler.resource_restrictions[key],
+                                start_index=i)
+                            if resource_i is None:
+                                continue
+                            else:
+                                i = resource_i
```

Comment:
Maybe this logic should just return the worker to steal from, rather than return an index into the idle collection.

Reply:
I tried to be minimally intrusive and stick to increasing the starting point for worker search...